waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/mapping/service/impl/PollingFederationServiceTest.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (C) 2016-2019 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.mapping.service.impl;
+
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.google.common.collect.Lists;
+
+import com.hotels.bdp.waggledance.api.model.AbstractMetaStore;
+import com.hotels.bdp.waggledance.api.model.MetaStoreStatus;
+import com.hotels.bdp.waggledance.core.federation.service.PopulateStatusFederationService;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PollingFederationServiceTest {
+
+  @Mock
+  private PopulateStatusFederationService populateStatusFederationService;
+
+  @Test
+  public void pollNotifyOnStateChange() throws Exception {
+    PollingFederationService service = new PollingFederationService(populateStatusFederationService);
+
+    AbstractMetaStore primary = AbstractMetaStore.newPrimaryInstance("p", "uri");
+    AbstractMetaStore federate = AbstractMetaStore.newFederatedInstance("f", "uri");
+    primary.setStatus(MetaStoreStatus.AVAILABLE);
+    federate.setStatus(MetaStoreStatus.AVAILABLE);
+
+    List<AbstractMetaStore> metastores = Lists.newArrayList(primary, federate);
+    when(populateStatusFederationService.getAll()).thenReturn(metastores);
+
+    // first time
+    service.poll();
+    verify(populateStatusFederationService, never()).update(primary, primary);
+    verify(populateStatusFederationService, never()).update(federate, federate);
+
+    // federated flipped status
+    federate.setStatus(MetaStoreStatus.UNAVAILABLE);
+    service.poll();
+
```

Comment:
I feel like the verify() should happen after you have polled and once more after the next status change. It reads better and easier to understand the test.

Otherwise, you can remove the service.poll() on line 61 and write another test to check what happens if the status changes from A to U and back again to A within 5 minutes and polling service doesn't see any change.

Reply:
I agree, i find its easier to read a long test like this if the expected behaviour is explicitly verified as you go through. What do others think?

Reply:
So would it be easier to read if we have the following tests:
* one metastore changes the status once and we verify that update was called after `service.poll()`
* one primary and one federated metastore change status at the same time (one from `UNAVAILABLE` to `AVAILABLE` and the other one, the other way around) and we verify that update is called for both after `service.poll()`
* one metastore changes status like this `AVAILABLE` &rightarrow; `UNAVAILABLE` &rightarrow; `AVAILABLE` with no `service.poll()` and we verify that there were no interactions

?

Reply:
I don't think it is necessary to split this into different tests. Would be easier to read if we verify what updates are done directly after the polls, including verifying that no updates are done to primary on the second poll, and not verifying twice at the end.

If you want to split it up, I would say keep the test exactly as it is, essentially:
- One poll, verify no updates are made
- Poll, don't change statuses, poll again and verify that no updates are made
- Poll, do change statuses, poll again and verify that updates are made

Reply:
Done :) I kept them in the same test.

Reply:
@max-jacobs I had what you described but changed it to do the last verify at the end: `verify(populateStatusFederationService, times(2)).update(federate, federate);` 
Because that last verify verifies everything so it as the times(2).
Which then looks a bit weird because it seems that the last poll does the update twice. 
With the amount of comments in the test I definitely think it should be rewritten. It's obviously not clear to everyone :)